### PR TITLE
[JENKINS-50518] Only run upload/download in the agent

### PR DESCRIPTION
**Pull request by carlossg**
Using presigned URLs

[JENKINS-50518](https://issues.jenkins-ci.org/browse/JENKINS-50518)

### CODE REVIEW

File: src/test/java/io/jenkins/plugins/artifact_manager_s3/JCloudsArtifactManagerTest.java [hunk 1]
```diff
@@ -67,8 +77,31 @@ public static void live() {
...
+    private static class ArtifactManagerFactoryForTesting extends JCloudsArtifactManagerFactory {
+        private String uuid;
+
+        public ArtifactManagerFactoryForTesting(String uuid) {
+            this.uuid = uuid;
+        }
+
+        @Override
+        public ArtifactManager managerFor(Run<?, ?> build) {
+            // use a different dir for each test
```
[the remark this answers is not in the document]
Comment (carlossg) [reply src/test/java/io/jenkins/plugins/artifact_manager_s3/JCloudsArtifactManagerTest.java]: that would still conflict if cleanup fails or if tests are run in parallel against same bucket

File: src/main/java/io/jenkins/plugins/artifact_manager_s3/S3BlobStore.java [hunk 2]
```diff
@@ -66,7 +74,8 @@ public BlobStoreContext getContext() throws IOException {
         LOGGER.log(Level.FINEST, "Building context for {0}", id());
         ProviderRegistry.registerProvider(AWSS3ProviderMetadata.builder().build());
         try {
-            return ContextBuilder.newBuilder(id()).credentialsSupplier(getCredentialsSupplier()).buildView(BlobStoreContext.class);
+            return ContextBuilder.newBuilder(id()).credentialsSupplier(getCredentialsSupplier())
+                    .buildView(BlobStoreContext.class);
```
[the remark this answers is not in the document]
Comment (carlossg) [reply src/main/java/io/jenkins/plugins/artifact_manager_s3/S3BlobStore.java]: unintended

File: src/main/java/io/jenkins/plugins/artifact_manager_s3/S3BlobStore.java [hunk 3]
```diff
@@ -98,23 +107,44 @@ public Credentials get() {
...
-    public URI toURI(String container, String key) {
+    public URI toURI(@NonNull String container, @NonNull String key) {
+        assert container != null;
+        assert key != null;
         try {
-            return new URI(String.format("https://%s.s3.amazonaws.com/%s", container, key));
-        } catch (URISyntaxException e) {
+            // TODO proper encoding
+            return new URI(String.format("https://%s.s3.amazonaws.com/%s", container,
+                    URLEncoder.encode(key, "UTF-8").replaceAll("%2F", "/").replaceAll("%3A", ":")));
```
[the remark this answers is not in the document]
Comment (carlossg) [reply src/main/java/io/jenkins/plugins/artifact_manager_s3/S3BlobStore.java]: no, this fixes `JCloudsBlobStoreTest#toURI` weird chars test

File: src/test/java/io/jenkins/plugins/artifact_manager_s3/JCloudsBlobStoreTest.java [hunk 4]
```diff
@@ -197,9 +197,15 @@ public void open() throws Exception {
...
-        assertEquals(new URI(String.format("https://%s.s3.amazonaws.com/%s", getContainer(), getPrefix().replaceFirst("/$", ""))),
+        assertEquals(new URI(
+                String.format("https://%s.s3.amazonaws.com/%s", getContainer(), getPrefix().replaceFirst("/$", ""))),
                 subdir.toURI());
         assertEquals(new URI(String.format("https://%s.s3.amazonaws.com/%s", getContainer(), filePath)), vf.toURI());
+        // weird chars
+        assertEquals(
+                new URI(String.format("https://%s.s3.amazonaws.com/%s", getContainer(),
+                        "xxx%23%3F:%24%26%27%22%3C%3E%C4%8D%E0%A5%90")),
+                newJCloudsBlobStore("xxx#?:$&'\"<>čॐ").toURI());
```
[the remark this answers is not in the document]
Comment (carlossg) [reply src/test/java/io/jenkins/plugins/artifact_manager_s3/JCloudsBlobStoreTest.java]: not yet, there are errors from AWS API

File: src/test/java/io/jenkins/plugins/artifact_manager_s3/JCloudsArtifactManagerTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jglick) [reply src/test/java/io/jenkins/plugins/artifact_manager_s3/JCloudsArtifactManagerTest.java]: True. Maybe we can do both—use a randomized test dir, and delete it `@After`?

**Pull request merged**